bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -25,219 +24,396 @@ import jsky.app.ot.gemini.tpe.TpePWFSFeature
+
+  implicit val OrderObsKey: Order[ObsKey] =
+    Order.order((k0,k1) => Ordering.fromInt(k0.k.compareTo(k1.k)))
+}
+
+
+object BagsManager {
+  private val Log = Logger.getLogger(getClass.getName)
+
+  // Worker pool running the AGS search and the timer.
+  val NumWorkers = math.max(1, Runtime.getRuntime.availableProcessors - 1)
+  private val worker = new ScheduledThreadPoolExecutor(NumWorkers, new ThreadFactory() {
+    override def newThread(r: Runnable): Thread = {
+      val t = new Thread(r, "BagsManager - Worker")
+      t.setPriority(Thread.NORM_PRIORITY - 1)
+      t.setDaemon(true)
+      t
+    }
+  })
+  private implicit val executionContext = ExecutionContext.fromExecutor(worker)
+
+  // This is our mutable state.  It is only read/written by the Swing thread.
+  private var stateMap  = ==>>.empty[ProgKey, ObsKey ==>> BagsState]
+
+  // Handle a state machine transition.  Note we switch to the Swing thread
+  // here so that the calling thread continues immediately.  All updates are
+  // done from the Swing thread.
+  private def transition(obs: ISPObservation, update: BagsState => StateTransition): Unit = Swing.onEDT {
+    val progKey = ProgKey(obs)
+    val obsKey  = ObsKey(obs)
+
+    val stateTuple = for {
+      obsMap <- stateMap.lookup(progKey)
+      state  <- obsMap.lookup(obsKey)
+    } yield (obsMap, state)
+
+    stateTuple.foreach { case (obsMap, state) =>
+      val (newState, sideEffect) = update(state)
+      val newStatemap            = stateMap.insert(progKey, obsMap.insert(obsKey, newState))
+
+      val action = for {
+        _ <- IO { stateMap = newStatemap }
+        _ <- sideEffect
+      } yield ()
+
+      action.unsafePerformIO()
     }
   }
 
+  private def edited(obs: ISPObservation): Unit =
+    transition(obs, _.edit)
+
+  private def wakeUp(obs: ISPObservation): Unit =
+    transition(obs, _.wakeUp)
+
+  private def success(obs: ISPObservation, results: Option[AgsStrategy.Selection]): Unit =
+    transition(obs, _.succeed(results))
+
+  private def fail(obs: ISPObservation, why: String, delay: Long): Unit =
+    transition(obs, _.fail(why, delay))
+
+  /** Add a program to our watch list and attach listeners. Enqueue all
+    * observations for consideration for a BAGS lookup.
+    */
+  def watch(prog: ISPProgram): Unit = {
+    // Check that the program isn't being managed already. If it is, do nothing.
+    val progKey = ProgKey(prog)
+    if (stateMap.lookup(progKey).isEmpty) {
+      prog.addCompositeChangeListener(CompositePropertyChangeListener)
+
+      // Place all observations in the IdleState and record them in our stateMap.
+      val obsList = prog.getAllObservations.asScala.toList
+      val obsMap = ==>>.fromList(obsList.map(o => ObsKey(o) -> (IdleState(o, None): BagsState)))
+
+      stateMap = stateMap + (ProgKey(prog) -> obsMap)
+
+      // Now mark all the observations "edited" in order to kick off AGS searches
+      // and hash calculations.
+      obsList.foreach(edited)
```

Comment:
Good question, it is probably possible to do it differently.

`PendingState` requires that it receive a `wakeUp` event to transition either back to `IdleState` if nothing need be done or else to `RunningState` if an AGS lookup is needed.  So we'd be stuck in `PendingState` forever if there isn't a subsequent `wakeUp` event. When an `IdleState` is `edited`, the action associated with the transition is to schedule the `wakeUp`.

Perhaps it would be just as well to start in `PendingState` and explicitly call `wakeUp` instead but I preferred having only the timer calling `wakeUp`.